### PR TITLE
Multi-select case list: force auto select app configuration

### PR DESCRIPTION
**Pull request by stephherbers**
## Product Description

Ticket: [USH-2436](https://dimagi-dev.atlassian.net/browse/USH-2436)
This is split up into 2 PRs for the HQ side: [here](https://github.com/dimagi/commcare-hq/pull/32304)

There is currently no way to force the user to select all matches from their search results, which could result in dual admissions or duplicates being created. This creates an app configuration setting to force auto selection for multi-select case lists with a given max-selection value. NOTE: the max-select value applies to both manual and auto select for multi-select case lists.  

## Technical Summary

Screen Recording:
Screen Recording 2022-11-01 at 11 16 07 AM

Note that auto-select is dependent on it being a multi-select case list, so if multi-select is disabled, auto-select will also be disabled, but the max-select-value is dependent on multi-select but **not** on auto-select. 

## Feature Flag

MULTI_SELECT_CASE_LIST

### Automated test coverage

not for the app cpnfiguration

### QA Plan

This will go through QA as a whole along with the formplayer and xml changes
[ticket](https://dimagi-dev.atlassian.net/browse/QA-4555)

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html [hunk 1]
```diff
@@ -11,13 +19,29 @@ <h4 class="panel-title panel-title-nolink">
...
+            <label>
+              <input type="checkbox" id="auto-select-checkbox" data-bind="checked: autoSelectEnabled, disable: !multiSelectEnabled()"/>
+              {% trans "Enable auto select" %}
+            </label>
+          </div>
+          <div data-bind="visible: multiSelectEnabled()">
+            <label class="control-label {% css_label_class %}">
+                {% trans "Maximum selected value" %}
+                <span class="hq-help-template" data-title="{% trans 'Maximum Selected Value' %}"
+                      data-content="{% trans_html_attr 'Sets a maximum selected value for multi select case lists. If no value is inputted, the default value is used.' %}">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:39]: It'd be nice to say what the default value is, although it's a little bit annoying to need to pass the constant in here.

Although, if you did pass the constant into the template, you could also stick it in initial page data and then have javascript access the real value instead of having a magic number in two places.

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html [hunk 2]
```diff
@@ -11,13 +19,29 @@ <h4 class="panel-title panel-title-nolink">
     </div>
     <div class="panel-body">
       <div class="row">
-        <div class="col-sm-8">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:14]: Why is this removed?

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html [hunk 3]
```diff
@@ -3,6 +3,14 @@
 
 {% if request|toggle_enabled:"USH_CASE_LIST_MULTI_SELECT" %}
 
+	<script>
+		function reconcileCheckboxes() {
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:7]: This logic should be part of the initialization of `autoSelectEnabled` in `screen.js`

**Review by orangejenny — COMMENTED** [review]
Left a couple of specific comments, will reach out to you & Shubham about a higher level design question (whether or not this could apply to normal, single-select case lists).

File: corehq/apps/app_manager/static/app_manager/js/details/screen.js [hunk 4]
```diff
@@ -70,14 +70,20 @@ hqDefine("app_manager/js/details/screen", function () {
         });
         self.multiSelectEnabled = ko.observable(spec[self.columnKey].multi_select);
         self.multiSelectEnabled.subscribe(function () {
+            if (!self.multiSelectEnabled()) {
```
Comment (Jtang-1) [inline corehq/apps/app_manager/static/app_manager/js/details/screen.js:73]: Ethan's possible suggestion: `self.autoSelectEnabled(self.multiSelectEnabled() && self.autoSelectEnabled())`

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html (hunk 3 above)
Comment (stephherbers) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html]: https://github.com/dimagi/commcare-hq/pull/32303#:~:text=autoselect%20checkbox%20logic-,116b101,-allow%20auto_select%20and

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html (hunk 2 above)
Comment (stephherbers) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:14]: The spacing seemed off, especially with the new input and I couldn't figure out why it was there to start-- is that standard?

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html (hunk 1 above)
Comment (stephherbers) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:31]: I tried this and had little luck with getting it passed in. I agree it would be best to add this so unless this is blocking I will fiddle with it more after this PR is merged

Comment (orangejenny) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:31]: not blocking

File: corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_multi_select.html:14]: Yeah, it's standard to have a `row` div then contain a set of `col-X-Y` divs. Bootstrap has a max of 12 cols, but we often use 8 as the max so that the UI only takes up two-thirds of the screen.

I'm curious if it would work with putting each checkbox in its own row, probably with a col-sm-8 inside of it (either with the col-sm8 as a layer between the row and the checkbox, or adding the col-sm-8 to the checkbox div), but it's not worth spending time on if you've already futzed with it a bunch.

File: corehq/apps/app_manager/static/app_manager/js/details/screen.js [hunk 5]
```diff
@@ -70,13 +70,7 @@ hqDefine("app_manager/js/details/screen", function () {
...
         self.multiSelectEnabled = ko.observable(spec[self.columnKey].multi_select);
         self.multiSelectEnabled.subscribe(function () {
-            if (!self.multiSelectEnabled()) {
-                document.getElementById("auto-select-checkbox").checked = false;
-                self.autoSelectEnabled();
-            }
-            else {
-                document.getElementById("auto-select-checkbox").checked = self.autoSelectEnabled();
-            }
+            self.autoSelectEnabled(self.multiSelectEnabled() && self.autoSelectEnabled())
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/details/screen.js]: 👍

File: corehq/apps/app_manager/views/modules.py [hunk 6]
```diff
@@ -1128,7 +1133,10 @@ def edit_module_detail_screens(request, domain, app_id, module_unique_id):
         'long': params.get("long_custom_variables", None)
     }
     multi_select = params.get('multi_select', None)
-
+    auto_select = params.get('auto_select', None)
+    max_select_value = params.get('max_select_value', MULTI_SELECT_MAX_SELECT_VALUE)
```
Comment (orangejenny) [inline corehq/apps/app_manager/views/modules.py:1137]: Nitpick: couldn't this be `max_select_value = params.get('max_select_value') or MULTI_SELECT_MAX_SELECT_VALUE` and then drop the if statement?

**Review by orangejenny — APPROVED** [review]
Looks good to me. Left a couple new minor comments, but nothing is blocking.

File: corehq/apps/app_manager/views/modules.py (hunk 6 above)
Comment (stephherbers) [reply corehq/apps/app_manager/views/modules.py]: ah yes indeed thank you! [8aa96ca](https://github.com/dimagi/commcare-hq/pull/32303/commits/8aa96ca6d362536cc1b2cd91f57d02bd4fcc628a)

**Review by Jtang-1 — APPROVED** [review]
